src/coreclr/src/vm/genericdict.cpp

```diff
@@ -640,6 +768,105 @@ Dictionary::Restore(
+        GC_TRIGGERS;
+        PRECONDITION(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+        POSTCONDITION(CheckPointer(RETVAL));
+    }
+    CONTRACT_END;
+
+    Dictionary* pDictionary = pMD->GetMethodDictionary();
+
+#if !defined(CROSSGEN_COMPILE)
+    DictionaryLayout* pDictLayout = pMD->GetDictionaryLayout();
+    InstantiatedMethodDesc* pIMD = pMD->AsInstantiatedMethodDesc();
+    _ASSERTE(pDictLayout != NULL && pDictLayout->GetMaxSlots() > 0);
+
+    DWORD sizeFromDictLayout = DictionaryLayout::GetDictionarySizeFromLayout(pMD->GetNumGenericMethodArgs(), pDictLayout);
+    if (pIMD->GetDictionarySlotsSize() != sizeFromDictLayout)
+    {
+        _ASSERT(pIMD->GetDictionarySlotsSize() < sizeFromDictLayout);
+
+        pDictionary = (Dictionary*)(void*)pIMD->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(sizeFromDictLayout + sizeof(void*)));
+        ZeroMemory(pDictionary, sizeFromDictLayout + sizeof(void*));
+
+        // Slot[-1] points at previous dictionary to help with diagnostics when investigating crashes
+        *(byte**)pDictionary = (byte*)pIMD->m_pPerInstInfo.GetValue() + 1;
+        pDictionary++;
+
+        // Copy old dictionary entry contents
+        memcpy(pDictionary, (const void*)pIMD->m_pPerInstInfo.GetValue(), pIMD->GetDictionarySlotsSize());
+
+        ULONG_PTR* pSizeSlot = ((ULONG_PTR*)pDictionary) + pIMD->GetNumGenericMethodArgs();
+        *pSizeSlot = sizeFromDictLayout;
+
+        // Flush any write buffers before publishing the new dictionary contents
+        FlushProcessWriteBuffers();
+
+        // Publish the new dictionary slots to the type.
+        FastInterlockExchangePointer((TypeHandle**)pIMD->m_pPerInstInfo.GetValuePtr(), (TypeHandle*)pDictionary);
+    }
+#endif
+
+    RETURN pDictionary;
+}
+
+Dictionary* Dictionary::GetTypeDictionaryWithSizeCheck(MethodTable* pMT)
+{
+    CONTRACT(Dictionary*)
+    {
+       THROWS;
+       GC_TRIGGERS;
+       PRECONDITION(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+       POSTCONDITION(CheckPointer(RETVAL));
+    }
+    CONTRACT_END;
+
+    Dictionary* pDictionary = pMT->GetDictionary();
+
+#if !defined(CROSSGEN_COMPILE)
+    DictionaryLayout* pDictLayout = pMT->GetClass()->GetDictionaryLayout();
+    _ASSERTE(pDictLayout != NULL && pDictLayout->GetMaxSlots() > 0);
+
+    DWORD sizeFromDictLayout = DictionaryLayout::GetDictionarySizeFromLayout(pMT->GetNumGenericArgs(), pDictLayout);
+    if (pMT->GetDictionarySlotsSize() != sizeFromDictLayout)
+    {
+        _ASSERTE(pMT->GetDictionarySlotsSize() < sizeFromDictLayout);
+
+        // Expand type dictionary
+        pDictionary = (Dictionary*)(void*)pMT->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(sizeFromDictLayout + sizeof(void*)));
+        ZeroMemory(pDictionary, sizeFromDictLayout + sizeof(void*));
+
+        // Slot[-1] points at previous dictionary to help with diagnostics when investigating crashes
```

Comment:
Is this still needed with the simpler scheme?

Reply:
Shouldn't be really...

Reply:
I'll remove it